src/network/handler.rs

```diff
@@ -148,64 +154,115 @@ pub enum ClientErr {
+  pub token_index:  usize,
+  pub clients:      HashMap<usize, C>,
+  pub poll:         Poll,
   pub available_tokens: Vec<usize>
 }
 
 
-impl<Client: NetworkClient> ProustHandler<Client> {
-  fn accept(&mut self, event_loop: &mut EventLoop<Self>) {
-    if let Ok(Some(stream)) = self.listener.accept() {
+impl<C: Client> Server<C> {
+
+  pub fn new(addr: SocketAddr, poll: Poll) -> Self {
+    Server {
+      tcp_listener: TcpListener::bind(&addr.into()).unwrap(),
+      token_index: 1,
+      clients: HashMap::new(),
+      poll,
+      available_tokens: Vec::new()
+    }
+  }
+
+  pub fn run(&mut self) {
+    let mut events = Events::with_capacity(1024);
+
+    self.poll.register(&self.tcp_listener, SERVER, Ready::readable(), PollOpt::edge()).unwrap();
+
+    'main: loop {
+      self.poll.poll(&mut events, None).unwrap();
+
+      for event in events.iter() {
+        match event.token() {
+          SERVER => {
+            self.accept();
+          },
+          Token(t) => {
+            let kind = event.readiness();
+
+            if UnixReady::from(kind).is_hup() {
+              self.close(t);
+            }
+            else if UnixReady::from(kind).is_readable() {
+              self.client_read(t);
+            }
+            else {
+              self.client_write(t);
+            }
+          }
+          _ => unimplemented!(),
+        }
+      }
+    }
+  }
+
+  fn accept(&mut self) {
+    if let Ok((stream, addr)) = self.tcp_listener.accept() {
       let index = self.next_token();
-      println!("got client n°{:?}", index);
+      info!("got client n°{:?}", index);
       let token = Token(index);
-      event_loop.register_opt(&stream, token, Interest::all(), PollOpt::edge());
+
+      self.poll.register(&stream, token, Ready::all(), PollOpt::edge());
+
       self.clients.insert(index, Client::new(stream, index));
-    } else {
-      println!("invalid connection");
+    }
+    else {
+      error!("Invalid connection");
     }
   }
 
-  fn client_read(&mut self, event_loop: &mut EventLoop<Self>, tk: usize) {
-    //println!("client n°{:?} readable", tk);
-    if let Some(mut client) = self.clients.get_mut(&tk) {
+  fn next_token(&mut self) -> usize {
+    match self.available_tokens.pop() {
+      None => {
+        let index = self.token_index;
+        self.token_index += 1;
+        index
+      },
+      Some(index) => {
+        index
+      }
+    }
+  }
+
+  fn client_read(&mut self, tk: usize) {
+    let mut error = false;
 
+    if let Some(mut client) = self.clients.get_mut(&tk) {
       match client.state() {
         ClientState::Normal => {
-          //println!("state normal");
           match client.read_size() {
             Ok(size) => {
-              let mut buffer = ByteBuf::mut_with_capacity(size);
-
-              let capacity = buffer.remaining();  // actual buffer capacity may be higher
-              //println!("capacity: {}", capacity);
+              let mut buffer = BytesMut::with_capacity(size);
```

Comment:
it's probably a bad idea to allocate a `BytesMut` each time we want to read. I know the previous code did that, but it's not really a good thing. Maybe we'll put that in an issue to fix it later